yt/frontends/halo_catalog/io.py

```diff
@@ -104,9 +78,143 @@ def _count_particles(self, data_file):
         return {"halos": nhalos}
 
     def _identify_fields(self, data_file):
-        with h5py.File(data_file.filename, mode="r") as f:
-            fields = [("halos", field) for field in f]
+        with h5py.File(data_file.filename, "r") as f:
+            fields = [
+                ("halos", field) for field in f if not isinstance(f[field], h5py.Group)
+            ]
             units = dict(
                 [(("halos", field), parse_h5_attr(f[field], "units")) for field in f]
             )
         return fields, units
+
+
+class HaloDatasetIOHandler:
+    """
+    Base class for io handlers to load halo member particles.
+    """
+
+    def _read_particle_coords(self, chunks, ptf):
+        pass
+
+    def _read_particle_fields(self, dobj, ptf):
+        # separate member particle fields from scalar fields
+        scalar_fields = defaultdict(list)
+        member_fields = defaultdict(list)
+        for ptype, field_list in sorted(ptf.items()):
+            for field in field_list:
+                if (ptype, field) in self.ds.scalar_field_list:
+                    scalar_fields[ptype].append(field)
+                else:
+                    member_fields[ptype].append(field)
+
+        all_data = self._read_scalar_fields(dobj, scalar_fields)
+        all_data.update(self._read_member_fields(dobj, member_fields))
+
+        for field, field_data in all_data.items():
+            yield field, field_data
+
+    def _read_particle_selection(self, dobj, fields):
```

Comment:
This function is a direct copy of `IOHandlerGadgetFOFHaloHDF5._read_particle_selection`. The original lost version of this PR included a refactor which merged them. I intend to re-do that refactor at some point, but I think it would be easier to do after this PR has been merged. I appeal for clemency until then.

I'm also happy to (for now) replace this function definition explicitly with:
```
_read_particle_selection = IOHandlerGadgetFOFHaloHDF5._read_particle_selection
```
That would at least cut down on code to review and make it clear what work will be done in the future.